### PR TITLE
8320597: RSA signature verification fails on signed data that does not encode params correctly

### PR DESCRIPTION
**Pull request by alexeybakhtin**
Backport is not clean because encodeSignature and decodeSignature methods belong to the RSASignature class ( not RSAUtil ).
This is the reason for no changes in the RSAUtil class.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] [JDK-8320597](https://bugs.openjdk.org/browse/JDK-8320597) needs maintainer approval

### Issue
 * [JDK-8320597](https://bugs.openjdk.org/browse/JDK-8320597): RSA signature verification fails on signed data that does not encode params correctly (**Bug** - P3 - Approved)

### Reviewers
 * [Martin Balao](https://openjdk.org/census#mbalao) (@martinuy - **Reviewer**) ⚠️ Review applies to [253f1cc0](https://git.openjdk.org/jdk17u-dev/pull/2052/files/253f1cc0ee4463ebfeeb110202bfa2c339eef087)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk17u-dev.git pull/2052/head:pull/2052` \
`$ git checkout pull/2052`

Update a local copy of the PR: \
`$ git checkout pull/2052` \
`$ git pull https://git.openjdk.org/jdk17u-dev.git pull/2052/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2052`

View PR using the GUI difftool: \
`$ git pr show -t 2052`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk17u-dev/pull/2052.diff">https://git.openjdk.org/jdk17u-dev/pull/2052.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk17u-dev/pull/2052#issuecomment-1854876579)

### CODE REVIEW

File: src/java.base/share/classes/sun/security/rsa/RSASignature.java [hunk 1]
```diff
@@ -247,27 +256,19 @@ public static byte[] encodeSignature(ObjectIdentifier oid, byte[] digest)
...
-        }
-        if (algId.getEncodedParams() != null) {
-            throw new IOException("Unexpected AlgorithmId parameters");
-        }
-        byte[] digest = values[1].getOctetString();
-        return digest;
+        DerOutputStream out = new DerOutputStream();
+        DerOutputStream oidout = new DerOutputStream();
+        oidout.putOID(oid);
+        out.write(DerValue.tag_Sequence, oidout);
```
Comment (martinuy) [inline src/java.base/share/classes/sun/security/rsa/RSASignature.java:267]: Can we write these lines in the same way as in the 22 patch? (out.write(DerValue.tag_Sequence, new DerOutputStream().putOID(oid));)

Comment (alexeybakhtin) [reply src/java.base/share/classes/sun/security/rsa/RSASignature.java:267]: No, we can not. DerOutputStream::putOID() was changed in JDK20 to return itself : [JDK-8294987](https://github.com/openjdk/jdk/commit/823b53d8957ff06b66c35e676c25992e9707afe6#diff-cc247245c29a0ea3516813a4c50ac2b79c0d959c193082a1603c86d1a9dca6e3R319)
It is not backported to JDK17, so putOID returns void

Comment (martinuy) [reply src/java.base/share/classes/sun/security/rsa/RSASignature.java:267]: Ah, I see. Thanks.

**Review by martinuy — APPROVED** [review]
Thanks for this backport. Looks good to me. Before moving forward, I'd suggest to wait until the integration of 8302017 and making sure that there aren't any other changes that affect this backport.

**Comment by martinuy** [comment]
I requested jdk17u Maintainer approval.

**Pull request closed without merging**